File: demos/python_demos/image_retrieval_demo/image_retrieval_demo/visualizer.py
"""
 Copyright (c) 2019 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""

import numpy as np

import cv2

SIZE = 200
BLACK = (0, 0, 0)
GREEN = (0, 255, 0)
LWD = 2
TEXT_SIZE = 1.5
BORDER = 30


def add_top10_gallery_images(demo_image, impaths, distances, input_image, target_pos):
    """ Add top-10 most similar images from the gallery to demo image. """

    for index, impath in enumerate(impaths[:10]):
        image = cv2.imread(impath)
        image = cv2.resize(image, (SIZE, SIZE))

        h_shift = 2 * BORDER + input_image.shape[0] + (SIZE + BORDER) * (index // 5)
        w_shift = BORDER + (index % 5) * (SIZE + BORDER)

        demo_image[h_shift: h_shift + SIZE, w_shift: w_shift + SIZE] = image

        if distances is not None:
            cv2.putText(demo_image, '{}:{}'.format(index, int(distances[index] * 100) / 100),
                        (w_shift - BORDER, h_shift - 5), 1,
                        TEXT_SIZE, BLACK, LWD)
        else:
            cv2.putText(demo_image, '{}'.format(index), (w_shift - BORDER, h_shift - 5), 1,
                        TEXT_SIZE, BLACK, LWD)

        if target_pos == index:
            cv2.rectangle(demo_image, (w_shift, h_shift), (w_shift + SIZE, h_shift + SIZE),
                          GREEN, BORDER // 3)

    return demo_image


def visualize(image, target_pos, impaths, distances, input_size, compute_embedding_time,
              search_in_gallery_time, imshow_delay):
    """ Visualizes input video frame with detected ROI and
        top-10 most similar images from the gallery. """

    input_image = cv2.resize(image, (SIZE * 4, SIZE * 3))

    demo_image = np.ones(
        (input_image.shape[0] + SIZE * 2 + BORDER * 4, SIZE * 5 + BORDER * 6, 3),
        dtype=np.uint8) * 200

    demo_image[BORDER:BORDER + input_image.shape[0],
               BORDER:BORDER + input_image.shape[1]] = input_image

    if target_pos is not None and target_pos >= 0:
        target = cv2.imread(impaths[target_pos])
        target = cv2.resize(target, (SIZE, SIZE))

        target_loc = BORDER + input_image.shape[0] - SIZE, BORDER * 2 + input_image.shape[1]
        demo_image[target_loc[0]:target_loc[0] + SIZE, target_loc[1]:target_loc[1] + SIZE] = target

        cv2.putText(demo_image, 'Pattern:', (target_loc[1], target_loc[0] - 10), 1, TEXT_SIZE,
                    BLACK, LWD)
        cv2.putText(demo_image, 'Position: {}'.format(target_pos),
                    (BORDER * 2 + input_image.shape[1], BORDER * 2), 1, TEXT_SIZE, BLACK, LWD)

    cv2.putText(demo_image, 'Gallery size: {}'.format(len(impaths)),
                (BORDER * 2 + input_image.shape[1], BORDER * 2 + 30), 1, TEXT_SIZE, BLACK, LWD)
    if not np.isnan(compute_embedding_time):
        cv2.putText(demo_image,
                    'Embbeding (ms): {}'.format(int(compute_embedding_time * 10000) / 10.0),
                    (BORDER * 2 + input_image.shape[1], BORDER * 2 + 60), 1, TEXT_SIZE, BLACK, LWD)
    if not np.isnan(search_in_gallery_time):
        cv2.putText(demo_image,
                    'Gallery search (ms): {}'.format(int(search_in_gallery_time * 10000) / 10.0),
                    (BORDER * 2 + input_image.shape[1], BORDER * 2 + 90), 1, TEXT_SIZE, BLACK, LWD)

    cv2.putText(demo_image, 'Inp. res: {0}x{0}'.format(input_size),
                (BORDER * 2 + input_image.shape[1], BORDER * 2 + 120), 1, TEXT_SIZE, BLACK, LWD)

    demo_image = add_top10_gallery_images(demo_image, impaths, distances, input_image, target_pos)

    cv2.imshow('demo_image', demo_image)

    return cv2.waitKey(imshow_delay) & 0xff

File: demos/image_processing_demo/cpp/visualizer.cpp
/*
// Copyright (C) 2021-2024 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
*/

#include "visualizer.hpp"

#include <ctype.h>
#include <stddef.h>

#include <memory>
#include <stdexcept>
#include <string>

#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <models/results.h>
#include <pipelines/metadata.h>
#include <utils/ocv_common.hpp>

Visualizer::Visualizer(const std::string& type) {
    if (type == "sr")
        winName = "Image Processing Demo - Super Resolution (press A for help)";
    else if (type == "jr")
        winName = "Image Processing Demo - JPEG Restoration (press A for help)";
}

cv::Size Visualizer::getSize() {
    return resolution;
}

void Visualizer::handleKey(int key) {
    key = std::tolower(key);
    if (key == 'a') {
        isHelpShown = !isHelpShown;
    }
    if (cv::getWindowProperty(winName, 0) < 0) {
        mode = "result";
        disableTrackbar();
    }
    if (key == 'o') {
        mode = "orig";
        addTrackbar();
    }
    if (key == 'v') {
        mode = "diff";
        addTrackbar();
    }
    if (key == 'r') {
        mode = "result";
        cv::destroyWindow(winName);
        disableTrackbar();
    }
}

cv::Mat Visualizer::renderResultData(ImageResult result, cv::Size& newResolution) {
    if (!result.metaData) {
        throw std::invalid_argument("Renderer: metadata is null");
    }
    // Input image is stored inside metadata, as we put it there during submission stage
    inputImg = result.metaData->asRef<ImageMetaData>().img;
    if (inputImg.empty()) {
        throw std::invalid_argument("Renderer: image provided in metadata is empty");
    }

    if (!isResolutionSet) {
        setResolution(newResolution);
    }

    cv::resize(result.resultImage, result.resultImage, resolution);
    cv::resize(inputImg, inputImg, resolution);

    if (inputImg.channels() != result.resultImage.channels()) {
        cv::cvtColor(result.resultImage, resultImg, cv::COLOR_GRAY2BGR);
    } else {
        resultImg = result.resultImage;
    }
    changeDisplayImg();
    return displayImg;
}

void Visualizer::show(cv::Mat img) {
    if (img.empty()) {
        changeDisplayImg();
        img = displayImg;
    }

    if (isHelpShown) {
        int pad = 10;
        int margin = 60;
        int baseline = 0;
        int lineH = cv::getTextSize(helpMessage[0], cv::FONT_HERSHEY_COMPLEX_SMALL, 0.75, 1, &baseline).height + pad;
        for (size_t i = 0; i < 4; ++i) {
            putHighlightedText(img,
                               helpMessage[i],
                               cv::Point(pad, margin + baseline + (i + 1) * lineH),
                               cv::FONT_HERSHEY_COMPLEX,
                               0.65,
                               cv::Scalar(255, 0, 0),
                               2);
        }
    }

    cv::imshow(winName, img);
}

void Visualizer::changeDisplayImg() {
    displayImg = resultImg.clone();
    if (mode == "orig") {
        inputImg(cv::Rect(0, 0, slider, inputImg.rows)).copyTo(displayImg(cv::Rect(0, 0, slider, resultImg.rows)));
        markImage(displayImg, {"O", "R"}, static_cast<float>(slider) / resolution.width);
        drawSweepLine(displayImg);
    } else if (mode == "result") {
        markImage(displayImg, {"R", ""}, 1);
    } else if (mode == "diff") {
        cv::Mat diffImg;
        cv::absdiff(inputImg, resultImg, diffImg);
        double min, max;
        cv::minMaxLoc(diffImg, &min, &max);
        double scale = 255.0 / (max - min);
        diffImg = (diffImg - min) * scale;
        diffImg(cv::Rect(0, 0, slider, resultImg.rows)).copyTo(displayImg(cv::Rect(0, 0, slider, displayImg.rows)));
        markImage(displayImg, {"D", "R"}, static_cast<float>(slider) / resolution.width);
        drawSweepLine(displayImg);
    }
}

void Visualizer::markImage(cv::Mat& image, const std::pair<std::string, std::string>& marks, float alpha) {
    int pad = 25;
    std::pair<float, float> positions(static_cast<float>(image.cols) * alpha / 2.0f,
                                      static_cast<float>(image.cols) * (1 + alpha) / 2.0f);
    putHighlightedText(image,
                       marks.first,
                       cv::Point(static_cast<int>(positions.first) - pad, 25),
                       cv::FONT_HERSHEY_COMPLEX,
                       1,
                       cv::Scalar(0, 0, 255),
                       2);
    putHighlightedText(image,
                       marks.second,
                       cv::Point(static_cast<int>(positions.second), 25),
                       cv::FONT_HERSHEY_COMPLEX,
                       1,
                       cv::Scalar(0, 0, 255),
                       2);
}

void Visualizer::drawSweepLine(cv::Mat& image) {
    cv::line(image, cv::Point(slider, 0), cv::Point(slider, image.rows), cv::Scalar(0, 255, 0), 2);
}

void Visualizer::setResolution(cv::Size& newResolution) {
    resolution = newResolution;
    isResolutionSet = true;
}

void Visualizer::addTrackbar() {
    if (!isTrackbarShown) {
        cv::createTrackbar(trackbarName, winName, &slider, resolution.width);
        cv::setTrackbarMin(trackbarName, winName, 1);
        isTrackbarShown = true;
    }
}

void Visualizer::disableTrackbar() {
    isTrackbarShown = false;
}
